### PR TITLE
#61: [Persist emails before send] dispatch on delivery type (closes #61)

### LINKED ISSUE #61: [Persist emails before send] dispatch on delivery type
← #59

## requirements
- The user should be allowed to choose the delivery type (`AtLeastOnce`, `AtMostOnce`) when instantiating mailo

## specs
{optional: describe technical specs to implement this feature, if not obvious}

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by calippo**
Closes #61

### tests performed

> A Test Plan is used to show what you tested to make sure your code works fine. You should write here all that you did to test, and provide some results of your testing.

> These results can be screenshots, query results, or even just pointers to unit tests that successfully passed.

### tests not performed (domain coverage)

> At times not everything can be tested, and writing what hasn't been tested is just as important as writing what has been tested.

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.

### CODE REVIEW

**Comment by calippo** [comment]
@bytecodeguru you can safely ignore Codacy for the moment. 
We discussed this in person. No need to review the details, I need a validation of the approach.

File: src/main/scala/mailo/models.scala [hunk 1]
```diff
@@ -2,17 +2,26 @@ package mailo
 
 import akka.http.scaladsl.model.ContentType
 
+import io.buildo.enumero.annotations.enum
+
+@enum trait DeliverySemantic {
```
Comment (bytecodeguru) [inline src/main/scala/mailo/models.scala]: What about `DeliveryGuarantee`?

File: src/main/scala/mailo/package.scala [hunk 2]
```diff
@@ -1,135 +1,26 @@
...
-    bcc: Option[String] = None,
-    subject: String,
-    templateName: String,
-    params: Map[String, String],
-    attachments: List[Attachment] = Nil,
-    tags: List[String] = Nil
-  ): Future[Either[MailError, MailResponse]] =
-    sendinblueS3Mailo.send(to, from, cc, bcc, subject, templateName, params, attachments, tags)
+trait Mailo {
+  def send(mail: Mail): Future[Either[MailError, MailResult]]
```
Comment (bytecodeguru) [inline src/main/scala/mailo/package.scala:10]: ❤️

File: src/main/scala/mailo/package.scala [hunk 3]
```diff
@@ -1,135 +1,26 @@
...
+    mailData: MailData,
+    mailClient: MailClient,
+    deliverySemantic: DeliverySemantic = DeliverySemantic.AtMostOnce
+  )(
+    implicit
+    ec: ExecutionContext,
+    conf: Config = ConfigFactory.load()
+  ) = deliverySemantic match {
+    case DeliverySemantic.AtMostOnce => new AtMostOnceMailo(mailData, mailClient)
+    case DeliverySemantic.AtLeastOnce => new AtLeastOnceMailo(mailData, mailClient)
```
Comment (bytecodeguru) [inline src/main/scala/mailo/package.scala]: An alternative api design would be to allow the client to specify the delivery guarantee per message:

```scala
trait Mailo {
  def send(mail: Mail, deliveryGuarantee: DeliveryGuarantee = AtMostOnce): Future[Either[MailError, MailResult]]
}
```

But it is also possible with your proposal: just create and pass around both mailo instances. Let's stick with it for now.

**Review by bytecodeguru — COMMENTED** [review]
Nice refactor ❤️
I've left some inline comments.

File: src/main/scala/mailo/models.scala (hunk 1 above)
Comment (calippo) [reply src/main/scala/mailo/models.scala]: :+1:

**Pull request merged**